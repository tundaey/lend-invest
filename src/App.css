.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}

.Investments-title {
  margin-top: 0.5em
}

.invested-row {
  background-color: #f9f9f9
}
.invested-badge{
  background-color: green;
  display: inline-block;
  min-width: 10px;
  padding: 3px 7px;
  font-size: 12px;
  font-weight: 700;
  line-height: 1;
  color: #fff;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  border-radius: 10px;
}
