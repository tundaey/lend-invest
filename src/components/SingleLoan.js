import React from 'react';
import {Badge} from 'react-bootstrap'

const SingleLoan = (props) => (
    <tr className={props.loan.invested? 'invested-row ': ''}
        onClick={(event)=> props.onClick(event, props.loan)}>
        <td>{props.loan.id}</td>
        <td>{props.loan.title}</td>
        <td>{props.loan.tranche}</td>
        <td>{props.loan.available}</td>
        <td>{props.loan.annualised_return}</td>
        <td>{props.loan.term_remaining}</td>
        <td>{props.loan.ltv}</td>
        <td>{props.loan.amount}</td>
        <td><Badge bsClass={props.loan.invested? 'invested-badge': 'badge'}>{props.loan.invested? 'Invested': 'Eligible'}</Badge></td>
    </tr>
)

export default SingleLoan