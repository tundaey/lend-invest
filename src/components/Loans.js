import React from 'react';
import { Table} from 'react-bootstrap'
import SingleLoan from './SingleLoan'

const Loans = (props)=> (
    <div>
        <Table bordered condensed hover responsive>
            <thead>
                <tr>
                    <th>#</th>
                    <th>Title</th>
                    <th>Tranche</th>
                    <th>Available</th>
                    <th>Annualised Return</th>
                    <th>Term remaining</th>
                    <th>Ltv</th>
                    <th>Amount</th>
                </tr>
            </thead>
            <tbody>
                {props.loans.map((loan)=> <SingleLoan onClick={props.openModal} key={loan.id} loan={loan}/>)}
            </tbody>
        </Table>
    </div>
)

export default Loans