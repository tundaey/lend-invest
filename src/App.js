import React, { Component } from 'react';
import './App.css';
import Loans from './components/Loans'
import {loans} from './current-loans.json'
import NavBar from './components/Navbar'
import InvestForm from './components/InvestForm'

class App extends Component {
  constructor(){
    super()

    this.state = {
      title: 'LendInvest',
      isModalOpen: false,
      loans: [],
      selectedLoan: {},
      possibleInvestments: []
    }
  }

  componentDidMount(){
    this.setState({loans: loans, possibleInvestments: loans})
  }

  render() {
    return (
      <div>
        <NavBar title={this.state.title}/>
        <div className="container">
          <div className="row">
            <div className="col-md-8">
              <br/>
              <div className="col-md-6"><h1>All Current Loans</h1></div>
              <div className="col-md-6 Investments-title">
                <h3>Possible Investments: <span>{this.state.possibleInvestments.length}</span></h3>
                </div>
            </div>
            <div className="col-md-11">
              <Loans 
                openModal={this.openModal}
                isModalOpen={this.state.isModalOpen} 
                closeModal={this.closeModal} 
                loans={this.state.loans}/>

              {this.state.isModalOpen 
                ? <InvestForm 
                closeModal={this.closeModal}
                processInvestment={this.processInvestment}
                selectedLoan={this.state.selectedLoan} 
                show={this.state.isModalOpen}/>
              : null}
            </div>
          </div>
        </div>
      </div>
    );
  }

  closeModal = (event)=>{
    event.preventDefault()
    this.setState({isModalOpen: false, selectedLoan: {}})
  }

  openModal = (event, loan)=>{
    event.preventDefault()
    this.setState({isModalOpen: true, selectedLoan: loan})
  }

  processInvestment = (loan, investment)=>{
    const available = parseFloat(loan.available.replace(/,/g, ''));
    const newAvailableAmount  = available - investment
    loan['available']= newAvailableAmount.toLocaleString('en')
    const newLoans = this.state.loans.map((l)=> {
      if(l.id === loan.id){
        loan['invested'] = true
        return loan
      }
      return l
    })
    
    this.updatePossibleInvestments(loan)
    this.setState({isModalOpen: false, loans: newLoans})
  }

  updatePossibleInvestments = (loan)=>{
    const investments = this.state.possibleInvestments
    const invested = investments.indexOf(loan)
    if(invested > -1){
      investments.splice(invested, 1)
      this.setState({possibleInvestments: investments})
    }
  }


}

export default App;
