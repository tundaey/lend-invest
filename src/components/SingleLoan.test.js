import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import {spy} from 'sinon'

import {loans} from '../current-loans.json'
import SingleLoan from './SingleLoan'
import App from '../App'
import Loans from './Loans'

test('Ensure SingleLoan component renders correctly with the details of the loan', ()=> {
    const onClick = jest.fn()
    const wrapper = shallow(<SingleLoan onClick={onClick} loan={loans[0]}/>)
    const element = wrapper.find('tr')
    expect(element.length).toBe(1)
    expect(element.get(0).props.children.length).toBe(9)
    expect(element.get(0).props.children[0].props.children).toBe(loans[0].id)
    expect(element.get(0).props.children[1].props.children).toBe(loans[0].title)
    expect(element.get(0).props.children[2].props.children).toBe(loans[0].tranche)
    expect(element.get(0).props.children[3].props.children).toBe(loans[0].available)
    expect(element.get(0).props.children[4].props.children).toBe(loans[0].annualised_return)
    expect(element.get(0).props.children[5].props.children).toBe(loans[0].term_remaining)
    expect(element.get(0).props.children[6].props.children).toBe(loans[0].ltv)
    expect(element.get(0).props.children[7].props.children).toBe(loans[0].amount)
})

test('When User clicks on a loan, the form should appear.', ()=> {
    const AppComponent = mount(<App/>)
    const spy = jest.spyOn(AppComponent.instance(), 'openModal');

    const wrapper = mount(<SingleLoan onClick={spy} key={loans[0].id}  loan={loans[0]}/>)
    wrapper.find('tr').simulate('click')
    const loan = loans[0]

    expect(AppComponent.instance().state.isModalOpen).toBe(true)
    expect(AppComponent.instance().state.selectedLoan).toBe(loan)
})

test('SingleLoan component renders a snapshot properly', ()=>{
    const spy = jest.fn()
    const tree = renderer.create(<SingleLoan onClick={spy} loan={loans[0]} />).toJSON();
    expect(tree).toMatchSnapshot(); 
})
