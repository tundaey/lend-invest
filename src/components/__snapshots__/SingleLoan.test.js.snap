// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleLoan component renders a snapshot properly 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td>
    1
  </td>
  <td>
    Voluptate et sed tempora qui quisquam.
  </td>
  <td>
    A
  </td>
  <td>
    11,959
  </td>
  <td>
    8.60
  </td>
  <td>
    864000
  </td>
  <td>
    48.80
  </td>
  <td>
    85,754
  </td>
  <td>
    <span
      className="badge"
    >
      Eligible
    </span>
  </td>
</tr>
`;
